protos/lobby.proto

```diff
@@ -0,0 +1,105 @@
+import "game_info.proto"
+
+package papan;
+
+message RegisterUpdate {
+  oneof update {
+    GameType type_added = 1;
+    /* We assume that a server will not concurrently add two game types of the
+       same name. If one attempts to do so, the second one should be ignored */
+    string type_removed = 2;
+  }
+}
+
+message RegisterServerResponse {
+}
+
+message CreateGameLobby {
+  uint32 game_type_id = 1;
+  repeated map<string, string> exclusive_options = 2;
+  repeated map<string, OptionValues> group_options = 3;
+}
+
+message LobbyAction {
+  CreateGameLobby create = 1;
+}
+
+message GameInstance {
+  uint32 game_type_id = 1;
+  uint32 current_players = 2;
+  uint32 game_lobby_id = 3;
+  map<string, string> exclusive_options = 4;
+  repeated map<string, OptionValues> group_options = 3;
+}
+
+message LobbyUpdate {
+  oneof update {
+    // A server registered a new game types.
+    GameType type_added = 1;
+    // A server unregistered a game type, or disconnected
+    uint32 type_id_removed = 2;
+    // A player created or joined a game lobby
+    GameInstance instance_updated = 3;
+    // A game lobby became unavailable
+    uint32 instance_removed = 4;
+    // The player making this call created a game lobby
+    GameInstance instace_created = 5;
+  }
+}
+
+enum GameLobbyStateChange {
+  // When all players are ready
+  SET_READY = 0;
+  UNSET_READY = 1;
+  LEAVE = 2;
+}
+
+message GameLobbyAction {
+  oneof action {
+    string send_chat_message = 1;
+    GameLobbyStateChange finish = 2;
+    // Only the host can do this
+    string kick_player_id = 3;
+  }
+}
+
+message Player {
+  /* I expect ID to be specific to a single connection to the lobby by that
+     player */
+  string player_id = 1;
+  string player_name = 2;
+}
+
+message GameLobbyChatMessage {
+  string sender_player_id = 1;
+  string message_text = 2;
+}
+
+message GameInfo {
+  /* Both of these will be passed in the metadata of the RPC to the game server
+     that joins the game */
+  string game_instance_id = 1;
+  string game_instance_player_id = 2;
+}
+
+message GameLobbyUpdate {
+  oneof update {
+    GameLobbyChatMessage new_chat_message = 1;
+    Player player_joined = 2;
+    string player_left = 3;
```

Comment:
This looks odd at first: we use "Player" for "joined", but a "string" for "left". Took a few seconds for my lightbulb to go off.